enhancements/cloud-integration/openstack/failure-domains.md

```diff
@@ -0,0 +1,715 @@
+
+`StorageAvailabilityZone` will set the Cinder availability zone of any volume attached during initial machine creation. This is currently limited to a root volume, but may be extended to include additional volumes in the future. It will be an error to specify [`Zone`](https://github.com/openshift/api/blob/54592eea55395af31e82fa9a605eb45523115454/machine/v1alpha1/types_openstack.go#L358-L360) on a volume attached to a Machine with a failure domain that specifies `StorageAvailabilityZone`.
+
+`Ports` will be used to define a set of port to be attached to all Machines in a failure domain. They use the existing [`PortOpts`](https://github.com/openshift/api/blob/54592eea55395af31e82fa9a605eb45523115454/machine/v1alpha1/types_openstack.go#L283-L333) struct to define ports with all currently supported properties and attachments with the same semantics. Note specifically that the network or subnet a port will attach to must already exist. Unlike `ComputeAvailabilityZone` and `StorageAvailabilityZone` it will be allowed to specify this field on both the failure domain and the Machine. If both are defined they will be merged by putting the failure domain ports first.
+
+### Installer
+
+The installer is directly reponsible for initial creation of the control plane servers. It is only indirectly responsible for creation of workers through the creation of the appropriate manifests.
+
+We will add a `FailureDomains` field to [OpenStack MachinePool](https://github.com/openshift/installer/blob/dbbc890fa40bd49aa761fb03612415e964c9eaf2/pkg/types/openstack/machinepool.go#L5-L35). The semantics of this field will be identical to the semantics of the existing `Zones` field expanded to be complete failure domains. These semantics are described in the comment below. We will document `Zones` as being deprecated, but have no immediate plans to remove it.
+
+```go
+    // FailureDomains defines a set of failure domains to be used by this
+    // MachinePool. For the control plane, machines will be balanced between
+    // these failure domains. For workers the installer will create a seperate
+    // MachineSet for each defined failure domain. If empty, failure domains
+    // will not be used and a single worker MachineSet will be created.
+    //
+    // It is an error to define both zones and failureDomains.
+    // +optional
+    FailureDomains []OpenStackFailureDomain `json:"failureDomains,omitempty"`
+```
+
+The `MachinesSubnet` option in install-config can be used to define the default subnet where to attach servers that do not have a Failure domain, or servers in Failure domains that do not specify Ports. If no `MachinesSubnet` is specified, the Installer will create a subnet and use it as the `MachinesSubnet`.
```

Comment:
> The `MachinesSubnet` option in install-config can be used to define the default subnet where to attach servers that do not have a Failure domain, or servers in Failure domains that do not specify Ports.

So the only difference between `MachinesSubnet` and `Ports` on the failureDomains field is the fact that the Network of the Subnet on `MachinesSubnet` is not connected to any specific AZ, while it's connected on Ports? Plus ensuring all nodes have the same control plane subnets?

Reply:
There are a couple syntactic differences:
* `MachinesSubnet` is only one, while `Ports` is an array
* `MachinesSubnet` is a subnet, while you can use `Ports` to define a connection to a network if you like that.

The main difference though is that while `MachinesSubnet` is a cluster-wide property, `Ports` is a property of a failure domain. Which means that machines in separate failure domains can end up attached to separate networks or subnets.

> the Network of the Subnet on MachinesSubnet is not connected to any specific AZ

In both cases (`MachinesSubnet` and whatever a port in `Ports` is connecting to), we are not dealing with zones. The network or the subnet in question must already exist, so the user (or their OSP admin) will have created them in the Neutron zone of their liking prior to launching the Installer.

Did I answer the question?